Add config file support to @wordpress/env

noahtallen opened the pull request:
> _Note: This PR has been superseded by https://github.com/WordPress/gutenberg/pull/20002, so follow the information there instead._

## Description
With this PR, a user can specify a `wp-env.json` file with local dependencies for the tool to load. The goal is to make life easier for plugin devs who need to be working on multiple plugins and/or themes at the same time.

1. During the `wp-env start` command, look in the local directory for a `wp-env.json` file.
2. Parse the JSON out of this file into a `plugins` and `themes` array.
3. Use `detect-context` to create context objects for each dependency path. This is the same format used for the local context, which makes things easy. I also added the absolute path to the dependency and basname of the path to the context object so that they can be reused easily.
4. Map each dependency to the Docker config file as a volume.

Here is a JSON file I used to test this:

`wp-calypso/apps/full-site-editing/full-site-editing-plugin/wp-env.json`
```json
{
    "themes": [
        "../../../../themes/varia",
        "../../../../themes/maywood"
    ],
    "plugins": [
        "../../../../gutenberg"
    ]
}
```

That's pretty nasty with the relative paths, but I wanted all developers of the plugin to use it if the repos are in the same top-level directory. For me this is source/gutenberg, source/wp-calypso, etc. Home paths (i.e. `~/source`) don't work since the tilde is not supported in `fs`, and absolute paths wouldn't be flexible for other devs. I'm open to a better approach here if folks have ideas. :)

## How has this been tested?
I verified locally that the specified plugins and themes in the wp-env.json file were loaded into the local WordPress install created by `wp-env start`.

## Types of changes
New Feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

cc @epiqueras 

See here: https://github.com/WordPress/gutenberg/issues/17724#issuecomment-539717371

noahtallen committed:
Add path/pathName to Context object

We will use this info to build the docker config object

noahtallen committed:
Add a utility to resolve a list of dependencies.

It resolves the dependencies from the .wpenv file and returns them in the
same "context" format as detect-context.

noahtallen committed:
Use dependency config during wp-env start

Autopopulates the docker config with volume mappings for the dependency contexts.

noahtallen committed:
Add whitespace

epiqueras commented on packages/env/lib/create-docker-compose-config.js:
```diff
@@ -1,11 +1,17 @@
+	const dependencies = [ ...deps, currentContext ];
+	return dependencies.map( ( { path, pathName, type } ) => `      - ${ path }/:/var/www/html/wp-content/${ type }s/${ pathName }/\n` ).join( '' );
+}
+
```
This isn't really useful anywhere else, specially with the dependency on whitespace amount. I think we should just inline it.

epiqueras commented on packages/env/lib/detect-context.js:
```diff
@@ -13,21 +13,45 @@ const path = require( 'path' );
+ *
+ * @return {Context} The context of the directory. If a theme or plugin, the type property will contain 'theme' or 'plugin'.
+ */
+module.exports = async function detectContext( directoryPath = cwd ) {
```
```suggestion
module.exports = async function detectContext( directoryPath = process.cwd() ) {
```

epiqueras commented on packages/env/lib/detect-context.js:
```diff
@@ -13,21 +13,45 @@ const path = require( 'path' );
+ * @type {Object}
+ * @property {string} type
+ * @property {string} path
+ * @property {string} pathName
```
```suggestion
 * @property {string} pathBasename
```

`cwdName` made sense, because it was the name of the current working directory, but here it's not the path's name, it's the path's basename.

noahtallen replied:
Good call

epiqueras commented on packages/env/lib/detect-context.js:
```diff
@@ -13,21 +13,45 @@ const path = require( 'path' );
+/**
+ * Detects the context of a given path.
+ *
+ * @param {string} directoryPath The directory to detect. Should point to a directory, defaulting to the current directory.
```
We need to document default parameters.

https://jsdoc.app/tags-param.html#optional-parameters-and-default-values

epiqueras commented on packages/env/lib/env.js:
```diff
@@ -127,6 +129,7 @@ module.exports = {
 		await Promise.all( [
 			activateContext( context ),
 			activateContext( context, true ),
+			...dependencies.map( activateContext ),
```
We also need to do this after `clean` runs.

epiqueras commented on packages/env/lib/resolve-dependencies.js:
```diff
@@ -0,0 +1,44 @@
+ * @return {Array<detectContext.Context>} An array of dependencies in the context format.
+ */
+module.exports = async function resolveDependencies() {
+	const envFile = await readFile( './.wpenv' );
```
I think the name of the file should match the name of the package binary:

Similarly to `wp-scripts`, here it's:

`@wordpress/env: wp-env`.

So the config file should be `./.wp-env`.

epiqueras reviewed (changes requested):
Great work here, almost ready to merge!

noahtallen committed:
Cleanup variable and file names from feedback

- Rename pathName to pathBasename
- Use .wp-env instead of .wpenv
- Inline useless function
- Clarify default param in jsdoc for detectContext

noahtallen committed:
Activate dependencies on clean action

noahtallen commented:
Thanks @epiqueras! I resolved your feedback and added dependency activation to the clean command. Ready for another look now :)

youknowriad commented:
I might be a bit late here but I wonder about supporting the generic `.env` file (it's something a lot of tools and projects are adopting) instead of a custom `wpenv` config?

I believe we already make use of environment variables in some of these configs, so it seems like a natural fit?

epiqueras commented:
@noahtallen Thanks for the iteration, one last thing to fix.

@youknowriad Using environment variables for arrays is really hard to wrangle and is a ripe incompatibility and conflict area in different OSs. Furthermore, I wouldn't be surprised if WP requires us to add a lot more options down the road. What I do think we should do is add the `.json` file extension to make sure all editors pick up on the format out of the box. cc @noahtallen

noahtallen commented:
@epiqueras Something like `wp-env.json` instead of `.wp-env`?

noahtallen committed:
Don't activate deps on test environment clean

noahtallen commented:
I removed dependency activation for the test environment `clean` :)

epiqueras commented:
>@epiqueras Something like wp-env.json instead of .wp-env?

Exactly. What do you think? It seems like most tools are gravitating towards this now to be explicit. Just looking at other config files in this repo supports this.

noahtallen commented:
I have no strong feelings on the issue. I don't see a strong reason to use either, though the automatic json detection is nice.

Here's the wp-calypso repo as an example of a large JS app:
<img width="263" alt="Screen Shot 2019-10-29 at 11 34 48 AM" src="https://user-images.githubusercontent.com/6265975/67798178-44c31f80-fa40-11e9-81de-fcd8f10b73f6.png">

The `.*` format for config files seems pretty standard, but I don't think it really affects developers. They'll have to follow documentation either way :D

youknowriad commented:
docker-compose and docker are already relying on `.env` files and it's one of the underlying tools for  `@wordpress/env`. more and more tools are using this pattern.

I see two main pros:

 - no new format to learn, just variables names 
 - the growing list of config files at the root level of different projects makes me nervous :)

I agree that arrays are difficult but there are workarounds (comma separation...).
I have a preference from `.env` files but I respect your decisions here. It is also something we can  reconsider in the future.

epiqueras commented:
Docker uses YAML files.

`.env` files are for global environment variables not Docker configuration options.

We would also have to prefix every variable to avoid conflicts. It's just not practical.

epiqueras commented:
@noahtallen Let's go with JSON for the detection reasons.

noahtallen committed:
Use wp-env.json for config file

noahtallen commented:
done

epiqueras commented on packages/env/lib/resolve-dependencies.js:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Returns an array of dependencies to be mounted in the Docker image.
+ *
+ * Reads from the .wpenv file in the current directory and uses detect
```
```suggestion
 * Reads from the wp-env.json file in the current directory and uses detect
```

noahtallen committed:
Use correct filename in comment

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

noahtallen commented:
@epiqueras Feel free to merge :) (I don't have access)

epiqueras commented:
We need to update the README for this.

noahtallen commented:
@epiqueras true! Here it is: https://github.com/WordPress/gutenberg/pull/18643

epiqueras commented:
Thank you!

WraithKenny commented:
As a note on the `.env` discussion (not for code, but for documentation), it's worth it for the user to set the `COMPOSE_PROJECT_NAME=someunique` var so that unrelated projects don't conflict. (Docker falls back to use the folder name, which aren't always unique.)

mcsf commented:
This PR needs a post-merge update of its description to reflect the correct filename instead of `wpenv`.

noahtallen commented:
I've updated the description to use `wp-env.json` instead of `.wpenv` (the filename at the time of merge). I also added a note that this PR has now been superseded by https://github.com/WordPress/gutenberg/pull/20002, which changes it to `.wp-env.json`
